azure: document new bring-up

colemickens opened the pull request:
This requires merging this first: https://github.com/kubernetes/kubernetes/pull/23068

Preview: http://colemickens.github.io/docs/getting-started-guides/azure/

Few things:
1. Not sure if there should be a left-hand navigation of my new Azure page.
2. Not sure how to change the "turn key Azure" link in the main left navigation bar to point to this bringup instead of the js/coreos version.

cc: @davidopp

colemickens commented:
Other questions:

1. When would this go live at kubernetes.io? Is that after `1.2.0` is released, or are master docs shown on `kubernetes.io`?

2. How does `https://get.k8s.io | bash` handle cases like this where the provider won't exist in a released version? Is it okay to leave that instruction in the doc, or better to remove it and add after there has been a stable release?

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,104 @@
+
+## Prerequisites
+
+1. An Azure Subscription. If you don't already have one, you may create one on [azure.microsoft.com](https://azure.microsoft.com).
```
azure.microsoft.com/free is an alternative, too.

colemickens replied:
I'd thought about it, but the link in there now is less likely to change and has a huge green Free button in the middle of it.

paulmey replied:
Does Subscription need to be Capitalized?

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,104 @@
+```shell
+export KUBERNETES_PROVIDER=azure; curl -sS https://get.k8s.io | bash
+```
+
````
formatting of filenames in the note and the next paragraph are inconsistent.

colemickens replied:
You mean the `config-default.sh` specifically? I'm okay with that. The note is *very* informational, where as when it is emphasized later, it's specifically hinting at the user to take a look at it.

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,104 @@
+export AZURE_AUTH_METHOD="device"
+```
+
+By default, this will deploy a cluster with 4 Standard\_A1-sized VMs: one master node, three worker nodes. This process takes about 5 to 10 minutes. Once the cluster is up, connection information to the cluster will be displayed. Additionally, your `kubectl` configuration will be updated to know about this cluster and this new cluster will be set as the active context.
````
hmm maybe you can backtick the VM size as well for emphasis, that should eliminate the need for escaping.

colemickens replied:
:+1:

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,104 @@
+## Prerequisites
+
+1. An Azure Subscription. If you don't already have one, you may create one on [azure.microsoft.com](https://azure.microsoft.com).
+2. A Microsoft account with access to your own Subscription, or an organizational account with Owner access to the Subscription.
```
I wonder if mentioning MSA requirement is going to cause extra confusion. People who have an Azure subscription probably (>99%) already have a MS account and they are the owners of the subscription?

colemickens replied:
>People who have an Azure subscription probably (>99%) already have a MS account and they are the owners of the subscription?

I'm not sure if that assumption will always be true. I can change the phrasing though:

>An account with access to your subscription

Or is that too generic?

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,104 @@
+which in turn calls [cluster/azure/util.sh](http://releases.k8s.io/{{page.githubbranch}}/cluster/azure/util.sh)
+using [cluster/azure/config-default.sh](http://releases.k8s.io/{{page.githubbranch}}/cluster/azure/config-default.sh).
+
+You must set `AZURE_SUBSCRIPTION_ID` or you willl receive errors. It may be set in `cluster/azure/config-default.sh` or set as environment variables:
```
typo:willl

colemickens replied:
:+1:

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,104 @@
+### Notes
+
+1. The user account used for these operations must have Contributor access to the Subscription.
+2. You can find your Subscription ID in the Azure Portal. (All Resources -> Subscriptions)
```
Link to azure portal here.

Also `->` can be replaced with `&rarr;` (→)

colemickens replied:
Will do. (Do I have to html escape it in the markdown or is it safe to paste the character in?)

ahmetalpbalkan replied:
You can just copy paste the rendering above or type in HTML without escaping. Both are fine.

ahmetalpbalkan commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,104 @@
+* Please see the [azkube](https://github.com/colemickens/azkube) repository for more information about the deployment tool that manages the deployment.
+* Please see the [Kubernetes docs](/docs/) for more details on administering
+and using a Kubernetes cluster.
+
```
extra \n at the EOF

paulmey commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,103 @@
+
+### Notes
+
+1. The user account used for these operations must have Contributor access to the Subscription.
```
It says 'Owner' in the section at the top, but 'Contributor' here, do they apply to different sets of operations?

theobolo commented:
inside the Utils.sh in /cluster/azure/ there is a problem with the AZURE_DOCKER_REGISTRY and AZURE_DOCKER.

By default the utils.sh is looking to gcr.io/google_containers/"hyperkubeimage"
But the last hyperkube 1.1.2 release can find at gcr.io**/v2/**google_containers/"hyperkubeimage"

You can Add to set  

export  AZURE_DOCKER_REGISTRY="gcr.io/v2"
export  AZURE_DOCKER="google_containers"

Or just change the utils.sh

Cheers

colemickens commented:
Could use some guidance on this now. It looks like `1.2.0` was cut before my updates to the bring-up were merged (or it wasn't taken to the 1.2.0 release branch). That means that these docs are not accurate for the 1.2.0 release.

Should I redo this then to be accurate for the 1.2.0 release?

davidopp commented:
Uh oh, which PRs did we miss, Cole?

colemickens commented:
Hi @davidopp, it's not so much that a PR was missed as much as it was awkward timing of my initial PR and the 1.2.0 release and then this PR for the docs.

I have another change that should be pulled into the main repo: https://github.com/kubernetes/kubernetes/pull/23344

This PR, for the documentation, is now up to date as well. **But** it's only accurate after that pull request. I was looking for some guidance on how to handle documentation for `1.2.0`, but maybe it's best to just handle that one-off until `1.2.1` which might include my subsequent changes to the Azure bringup?

colemickens commented:
ping @davidopp

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+
+The Azure deployment process produces an output directory `cluster/azure/_deployments/${AZURE_DEPLOY_ID}`. In this directory you will find the PKI and SSH assets created for the cluster, as well as a script named `util.sh`. Here are some examples of its usage:
+
+```
````
50: ```shell

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+
+`kubectl proxy` is currently used to access to access deployed services.
+
+```
````
78: ```shell

johndmulhausen commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,154 @@
+
+#### Addon: SkyDNS
+
+You can deploy the [SkyDNS addon](https://github.com/kubernetes/kubernetes/tree/release-1.2/cluster/addons/dns):
```
URL: https://github.com/kubernetes/kubernetes/tree/{{page.githubbranch}}/cluster/addons/dns

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+
+You can deploy the [SkyDNS addon](https://github.com/kubernetes/kubernetes/tree/release-1.2/cluster/addons/dns):
+
+```
````
89: ```shell

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+
+This will deploy the [`kube-dashboard`](https://github.com/kubernetes/dashboard) addon:
+
+```
````
98: ```shell

johndmulhausen commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,154 @@
+
+#### Example: Guestbook
+
+This will deploy the [`guestbook example`](https://github.com/kubernetes/kubernetes/blob/release-1.2/examples/guestbook/README.md) (the all-in-one variant):
```
URL should be https://github.com/kubernetes/kubernetes/blob/{{page.githubbranch}}/examples/guestbook/README.md

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+
+This will deploy the [`guestbook example`](https://github.com/kubernetes/kubernetes/blob/release-1.2/examples/guestbook/README.md) (the all-in-one variant):
+
+```
````
108: ```shell

johndmulhausen commented on docs/getting-started-guides/azure.md:
````diff
@@ -0,0 +1,154 @@
+The `azkube` tool used internally during `kube-up` can also be used to scale your cluster.
+Here's an example of scaling a default deployment of 3 nodes to 10 nodes:
+
+```
````
120: ```shell

johndmulhausen commented on docs/getting-started-guides/index.md:
```diff
@@ -48,7 +48,7 @@ few commands, and have active community support.
 - [GCE](/docs/getting-started-guides/gce)
 - [AWS](/docs/getting-started-guides/aws)
-- [Azure](/docs/getting-started-guides/coreos/azure/)
+- [Azure](/docs/getting-started-guides/azure)
```
Hm, I think the original coreos/azure looks right?

colemickens replied:
CC: @davidopp 

We had discussed whether we wanted to replace the existing CoreOS/Azure or leave them side-by-side.

davidopp replied:
Yeah, let's not remove the other one.  But we should add some information so people can distinguish them. How about something like
```
[Azure](/docs/getting-started-guides/coreos/azure/) (Weave-based, contributed by WeaveWorks employees)
[Azure](/docs/getting-started-guides/azure/) (Flannel-based, contributed by Microsoft employee)
```
unless @colemickens doesn't want his affiliation mentioned.

johndmulhausen commented on docs/getting-started-guides/index.md:
```diff
@@ -115,6 +115,7 @@ GKE                  |              |        | GCE         | [docs](https://clou
 Vagrant              | Saltstack    | Fedora | flannel     | [docs](/docs/getting-started-guides/vagrant)                                | ['œ“][2]   | Project
 GCE                  | Saltstack    | Debian | GCE         | [docs](/docs/getting-started-guides/gce)                                    | ['œ“][1]   | Project
 Azure                | CoreOS       | CoreOS | Weave       | [docs](/docs/getting-started-guides/coreos/azure/)                    |          | Community ([@errordeveloper](https://github.com/errordeveloper), [@squillace](https://github.com/squillace), [@chanezon](https://github.com/chanezon), [@crossorigin](https://github.com/crossorigin))
+Azure                | CoreOS       | CoreOS | flannel     | [docs](/docs/getting-started-guides/azure)                    |          | Community ([@colemickens](https://github.com/colemickens))
```
Again, I think coreos/azure is the URL

colemickens replied:
This is a new bring-up document with a new link. It currently is meant to live alongside the coreos/azure guide.

johndmulhausen commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,154 @@
+## Further reading
+
+* Please see the [azkube](https://github.com/colemickens/azkube) repository for more information about the deployment tool that manages the deployment.
+* Please see the [Kubernetes docs](/docs/) for more details on administering
```
Probably don't need this bullet; we're already here :)

johndmulhausen commented on docs/getting-started-guides/azure.md:
```diff
@@ -0,0 +1,154 @@
+4. The `--node-size` used in the `scale` command must be the same size deployed initially or it will not have the desired effect.
+5. Cluster tear-down requires manual intervention, due to the fact that it deletes the entire resource group and someone else may have deployed other resources since the initial deployment. For this reason you must confirm the list of resources that are to be deleted. If you wish to skip it, you may set `AZURE_DOWN_SKIP_CONFIRM` to `true`. This will delete everything in the resource group that wa ...
+6. If you are deploying from a checkout of `kubernetes`, then you will need to take an additional step to ensure that a `hyperkube` image is available. You can set `AZURE_DOCKER_REGISTRY` and `AZURE_DOCKER_REPO` and the deployment will ensure that a hyperkube container is built and available in the specified Docker registry. That `hyperkube` image will then be used throughout the cluster for runn ...
+
```
Lose a blank line here :)

davidopp commented:
Sorry for dropping this, Cole. The problem is that under the new docs system (with the new docs repo kubernetes/kubernetes.github.io), for now, all docs changes show up immediately and are not versioned. So I think what you should do is
1) Submit documentation now that describes the 1.2.0 behavior
2) Tell me which PR #s you want in 1.2.1, so I can make sure they get in to 1.2.1
3) After 1.2.1 is cut, update this doc to add a "behavior in 1.2.1" section (but not replace the 1.2.0 behavior).

Eventually we will be back to having versioned documentation, but we don't right now.

colemickens commented:
Hi David, I've refactored the docs so that they are applicable to both 1.2.0 (+1.2.1) and 1.2.2 (assuming the other PR is merged before then). The only differences are in whether AZURE_TENANT_ID is required and that kube-system is automatically deployed in 1.2.2.

I've addressed the other comments also. PTAL and let me know what you think.

colemickens commented:
This has been rebased for 1.3.0.

johndmulhausen commented:
Alternatively, if this is 1.3-bound, you could open a PR against the release-1.3 branch in this repo and then we can merge it at any time. The 1.3 branch will be "promoted" to master at 1.3 release time. 

@colemickens @davidopp

davidopp commented:
I didn't read the doc super-carefully, but except for my one comment, this looks fine to me.

colemickens commented:
I made the adjustment as requested David, both bring-ups are listed now.

colemickens commented:
Just added a new commit that adds this to the sidebar as well. One thing I noticed, @davidopp , the sidebar doesn't appear on my documentation page. I'm not sure what needs to be added for it to show up.

davidopp commented:
@johndmulhausen do you know the answer to cole's question directly above?

davidopp commented:
LGTM

@johndmulhausen can you take a look and merge if you're satisfied?

johndmulhausen commented:
Merging as it's not 1.3-exclusive, my nits are long-addressed, and @davidopp feels good.
